### PR TITLE
Feature/cloud 385/mask cloud creds

### PR DESCRIPTION
**Pull request by joshua-stauffer**
Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
- expands PasswordMasker functionality to handle cloud tokens
- adds a new method to DataContextConfig: `to_sanitized_json_dict`, which uses the new PasswordMasker functionality
- updates `DataContext.list_datasources()` and `DataContext.list_stores()` to use the same functionality
- resolves CLOUD-385

After submitting your PR, CI checks will run and @ge-cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

**Comment by joshua-stauffer** [comment]
Looks like two tests are failing due to something regarding bigquery -- it might make sense to remove bigquery from the test data, since there isn't a password to be masked anyways.

File: great_expectations/data_context/types/base.py
Comment (alexsherstinsky) [inline great_expectations/data_context/types/base.py]: @joshua-stauffer Could we please pick a different string?  For example, "abc".  Thanks!

**Review by alexsherstinsky — CHANGES_REQUESTED** [review]
@joshua-stauffer requesting a small change -- otherwise, happy to approve -- thanks!

File: great_expectations/data_context/types/base.py
Comment (cdkini) [inline great_expectations/data_context/types/base.py:2096]: ❤️

File: great_expectations/data_context/util.py
Comment (cdkini) [inline great_expectations/data_context/util.py]: I think this file is great but is there a way we could possibly genericize this whole utility? I'm wary of having to constantly expand this class to fit different core GE objects. I think it's probably fine as is but what are your thoughts around a recursive method that anonymizes anything in a list of keywords?

Something like:
```python
SENSITIVE_STRINGS = set(...)

@classmethod
def sanitize(cls, config: dict) -> dict:
    if not isinstance(config, dict):
        raise

    config_copy = safe_deep_copy(config)
    
    def _recursive_cleaner_method(config: Any) -> None:
        if isinstance(config, dict):
            for key, val in config.items():
                if key in self.SENSITIVE_STRINGS and isinstance(val, str):
                    # obfuscate
                _recursive_cleaner_method(val)
        elif isinstance(config, list):
            for val in config:
                _recursive_cleaner_method(val)

    _recursive_cleaner_method(config_config) # Perform anonymization in place
    
    return config_copy
```

File: tests/data_context/test_data_context_utils.py
Comment (cdkini) [inline tests/data_context/test_data_context_utils.py]: Although this is definitely a unit test, I feel like it's pretty complicated and is testing a lot of different things. Could we simplify this at all? Maybe a snapshot test or splitting it up into a few different tests?

Having `try/except` and `if/else` for various asserts makes debugging this a little difficult imo.

**Review by cdkini — CHANGES_REQUESTED** [review]
Really nice work! Two minor comments:
* Suggestion about possible way to make this more generic (not blocking).
* Request to split up/simplify the unit tests if at all possible.

Let me know your thoughts and I'd be more than happy to approve!

File: tests/data_context/test_data_context_utils.py
Comment (joshua-stauffer) [reply tests/data_context/test_data_context_utils.py]: thanks for the feedback -- i'll break these cases out into separate tests.

File: great_expectations/data_context/types/base.py
Comment (joshua-stauffer) [reply great_expectations/data_context/types/base.py]: hi @alexsherstinsky -- sure, I'll update this to match `PasswordMasker.MASKED_PASSWORD_STRING`. Thanks for the feedback!

File: great_expectations/data_context/util.py
Comment (joshua-stauffer) [reply great_expectations/data_context/util.py]: thanks @cdkini, this is a great idea -- I had originally decided against this because there is different processing required depending on the key value, but i believe integrating our approaches works well.

File: tests/data_context/test_data_context_utils.py
Comment (joshua-stauffer) [reply tests/data_context/test_data_context_utils.py]: I've factored the tests out into their own functions, but have left the try/except, since it's the most straightforward way to iterate through unknown values and check their configuration. Let me know if this is easier to parse!

File: great_expectations/data_context/types/base.py
Comment (cdkini) [inline great_expectations/data_context/types/base.py]: Are either of these comments needed? For the deprecation, do we have a plan in place to remove that?

File: tests/data_context/test_data_context_utils.py
Comment (cdkini) [inline tests/data_context/test_data_context_utils.py]: I still am a bit iffy on this general pattern as it introduces some tricky conditional logic in a test, which should be as simple and straightforward as possible (imo).

I won't block on this but I'm definitely of the opinion that the fewer asserts and overall conditional branches, the easier the test is to read and maintain.

File: tests/data_context/test_data_context_utils.py
Comment (cdkini) [inline tests/data_context/test_data_context_utils.py:386]: Same general thought here! Could we just assert against a dict so we know the full expected vs actual?

File: tests/data_context/test_data_context_utils.py
Comment (cdkini) [inline tests/data_context/test_data_context_utils.py]: Typo in all docstrings. Please fix before merging 😄 

Nitpick - since the name of the test specifies the method in question, is this necessary? I think you could probably get by without it since these are very simple (and well written!) unit test.

**Review by cdkini — APPROVED** [review]
LGTM but I am still a little iffy on the following tests:
* test_sanitize_config_masks_execution_engine_connection_strings
* test_sanitize_config_masks_cloud_store_backend_access_tokens

I'd love it if we could simplify the logic there just to make it easier to debug and work with. That being said, I definitely won't block on that. Really nice work here, thanks so much for resolving my comments!

File: great_expectations/data_context/types/base.py
Comment (joshua-stauffer) [reply great_expectations/data_context/types/base.py]: good point -- I'll remove the access token comment, that was just a reminder for myself, but the account_id comment is outside of the scope of this PR, so I'll leave it. I believe it's scheduled to be removed on the next major release, but I don't have the latest context on that.

File: tests/data_context/test_data_context_utils.py
Comment (joshua-stauffer) [reply tests/data_context/test_data_context_utils.py]: I hear your point on the complexity of the test! I refactored the try/except to make it more straightforward to read. I personally find this approach to be preferable to a snapshot-style test since I see it as less fragile. From my perspective, this test shouldn't care about the rest of the contents of the config dict, only the fields under test.

**Review by alexsherstinsky — APPROVED** [review]
LGTM

**Pull request merged**
